setup.py

```diff
@@ -1,42 +1,151 @@
 import cellprofiler.utilities.version
 
+# Recipe needed to get real distutils if virtualenv.
+# Error message is "ImportError: cannot import name dist"
+# when running app.
+# See http://sourceforge.net/p/py2exe/mailman/attachment/47C45804.9030206@free.fr/1/
+#
+if hasattr(sys, 'real_prefix'):
+    # Running from a virtualenv
+    assert hasattr(distutils, 'distutils_path'), \
+        "Can't get real distutils path"
+    libdir = os.path.dirname(distutils.distutils_path)
+    sys.path.insert(0, libdir)
+    #
+    # Get the system "site" package, not the virtualenv one. This prevents
+    # site.virtual_install_main_packages from being called, resulting in
+    # "IOError: [Errno 2] No such file or directory: 'orig-prefix.txt'
+    #
+    del sys.modules["site"]
+    import site
+
+    assert not hasattr(site, "virtual_install_main_packages")
+
 setuptools.dist.Distribution({
     "setup_requires": [
         "clint",
+        "javabridge",
+        "matplotlib",
+        "numpy",
+        "pytest",
         "requests",
-        "pytest"
+        "scipy",
+        "pyzmq"
     ]
 })
 
+try:
+    import matplotlib
+    import numpy  # for proper discovery of its libraries by distutils
+    import scipy.sparse.csgraph._validation
+    import scipy.linalg
+    import zmq  # for proper discovery of its libraries by distutils
+    import zmq.libzmq
+except ImportError:
+    pass
+
+if sys.platform.startswith("win"):
+    import _winreg
+
+    try:
+        import py2exe
+
+        has_py2exe = True
+    except:
+        has_py2exe = False
+else:
+    has_py2exe = False
+
+#
+# Recipe for ZMQ
+#
+if sys.platform.startswith("win"):
+    #
+    # See http://www.py2exe.org/index.cgi/Py2exeAndzmq
+    # Recipe needed for py2exe to package libzmq.dll
+    os.environ["PATH"] += os.path.pathsep + os.path.split(zmq.__file__)[0]
+
 
 class Install(setuptools.command.install.install):
     def run(self):
+        self.run_command("build_version")
+        self.run_command("build_java_dependencies")
+        setuptools.command.install.install.run(self)
+
+
+class Develop(setuptools.command.develop.develop):
+    def install_for_development(self):
+        self.reinitialize_command("build_java_dependencies", inplace=1)
+        setuptools.command.develop.develop.install_for_development(self)
+
+
+class BuildVersion(setuptools.Command):
+    user_options = [
+        ("version", None, "CellProfiler semantic version")]
+
+    def initialize_options(self):
+        self.version = None
+
+    def finalize_options(self):
+        if self.version is None:
+            self.version = self.distribution.metadata.version
+
+    def run(self):
+        with open(os.path.join("cellprofiler", "frozen_version.py"),
+                  "w") as fd:
+            fd.write("version_string='%s'\n" %
+                     cellprofiler.utilities.version.version_string)
+            fd.write("dotted_version='%s'\n" % self.version)
+
+
+class BuildJavaDependencies(setuptools.Command):
```

Comment:
While this solves the “CellProfiler looks for prokaryote in cellprofiler/imagej/jars” problem, I don’t know this is this is the best solution. I suggest we modify CellProfiler to only look for site-packages/imagej/jars. I admit that this is mostly my problem—I should’ve broke python-imagej into a published package when I made the prokaryote changes.